Apps/Src/ReadCarCAN.c

```diff
@@ -67,98 +93,145 @@ static inline void chargingEnable(void) {
-            restartingArray = true;
+            // Wait to make sure precharge is finished and then restart array
+            OSTmrStart(&prechargeDlyTimer, &err);
             
-            //Turn on the array restart thread
-            OSTaskCreate(
-                &arrayTCB,
-                "Array Restarter",
-                &ArrayRestart,
-                NULL,
-                TASK_READ_CAR_CAN_PRIO,
-                arraySTK,
-                DEFAULT_STACK_SIZE/10,
-                DEFAULT_STACK_SIZE,
-                0,
-                1,
-                NULL,
-                OS_OPT_TASK_STK_CLR,
-                &err
-            );
-            assertOSError(OS_READ_CAN_LOC,err);   
         }
-    }
+    
 
     OSMutexPost(&arrayRestartMutex,
                 OS_OPT_NONE,
                 &err);
     assertOSError(OS_READ_CAN_LOC,err);
 }
 
+/**
+ * @brief Callback function for the precharge delay timer. Turns off precharge and restarts the array.
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+ * 
+*/
+static void arrayRestart(void *p_tmr, void *p_arg){
+
+    if(!regenEnable){    // If regen enable has been disabled during precharge, we don't want to turn on the main contactor immediately after
+        restartingArray = false;
+    }
+    else {
+       
+            Contactors_Set(ARRAY_CONTACTOR, ON, SET_CONTACTORS_NON_BLOCKING);
+    }
+
+        //Lights_Set(A_CNCTR, ON); Don't know where this function is. Does it even exist?
+
+        // done restarting the array
+        restartingArray = false;
+
+};
+
+/**
+ * @brief Disables charging when canWatchTimer hits 0 (when charge enable messages are missed)
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+*/
+void canWatchTimerCallback (void *p_tmr, void *p_arg){
+    OS_ERR err;
+
+    // mark regen as disabled
+    regenEnable = OFF;
+    
+    // Set fault bitmaps 
+    FaultBitmap |= FAULT_READBPS;
+    OSErrLocBitmap |= OS_READ_CAN_LOC;
+
+    // Signal fault state to kill contactors at its earliest convenience
+    OSSemPost(&FaultState_Sem4, OS_OPT_POST_1, &err);
+    assertOSError(OS_READ_CAN_LOC,err);
+
+}
+
+
+
 void Task_ReadCarCAN(void *p_arg)
 {
     OS_ERR err;
     uint8_t buffer[8]; // buffer for CAN message
     uint32_t canId;
-    CPU_TS ts;
 
     OSMutexCreate(&arrayRestartMutex, "array restart mutex", &err);
     assertOSError(OS_READ_CAN_LOC,err);
-    OSMutexCreate(&msg_rcv_mutex, "message received mutex", &err);
-    assertOSError(OS_READ_CAN_LOC,err);
+    OSSemCreate(&restartArraySem, "array restart semaphore", 0, &err);
+    assertOSError(OS_READ_CAN_LOC, err);
+    OSSemCreate(&canWatchDisableChargingSem, "disable charging semaphore", 0, &err);
+
+    // Calculate threshold value for the charge-enable saturation buffer based on buffer length
```

Comment:
No need for a `for` loop, you can just do `saturationThreshold = (SAT_BUF_LENGTH + 1) * SAT_BUF_LENGTH / 4`

Reply:
It might be better to initialize this value outside of the function since it's just a simple assignment.